src/components/input/input.ts

```diff
@@ -0,0 +1,108 @@
+ */
+const shrinkElasticTextArea = (textarea: HTMLTextAreaElement) => {
+  const minRows = Number(textarea.dataset.minRows);
+  let rows = Number(textarea.getAttribute('rows')) || minRows;
+  let isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+  // Shrink until we hit the minimum rows or start scrolling
+  for (let i = rows; i >= minRows; i--) {
+    rows = i;
+    textarea.setAttribute('rows', String(Math.max(rows, minRows)));
+    isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+    // If we're scrolling, then we shrank too much, so grow a bit and stop
+    if (isScrolling) {
+      growElasticTextArea(textarea);
+      break;
+    }
+  }
+};
+
+/**
+ * Update Elastic TextArea
+ *
+ * When a textarea is updated, compares the new value to the previous value
+ * (from a dataset attribute) to determine whether to grow or shrink.
+ *
+ * Note: Ideally we'd destructure `event.target` from the parameters and declare
+ * the type directly, since we don't care about the rest of the `event` object.
+ * Unfortunately, that caused some TS problems. If someone finds a solution,
+ * we should update this function to take `{target}` as the only parameter.
+ * @see https://stackoverflow.com/a/28900856/4898045
+ *
+ * @param {Event} event - the input event on the target `textarea`
+ */
+const updateElasticTextArea = (event: Event) => {
+  const textarea = event.target as HTMLTextAreaElement;
+  const lastValue = textarea.dataset.lastValue || '';
+  const newValue = textarea.value;
+
+  // Check whether to grow or shrink
+  if (newValue.length > lastValue.length) {
+    growElasticTextArea(textarea);
+  } else {
+    shrinkElasticTextArea(textarea);
+  }
+
+  // Save the new value for comparing
+  textarea.dataset.lastValue = newValue;
+};
+
+/**
+ * Create Elastic TextArea
+ *
+ * Given a textarea, adds dataset attributes for the existing value and the
+ * minimum number of rows, for use in the update function, which is added to
+ * an event listener. Returns an object containing the `destroy()` method to
+ * remove the event listener.
+ *
+ * @param {HTMLTextAreaElement} textarea - the target `textarea` element
+ * @returns {object} object containing a `destroy()` method
+ */
+export const createElasticTextArea = (textarea: HTMLTextAreaElement) => {
+  // Save the current value for comparing
+  textarea.dataset.lastValue = textarea.value || '';
+
+  // Set the minimum rows to the initial `rows` value or 2
+  textarea.dataset.minRows = textarea.getAttribute('rows') || '2';
+
+  // Add an event listener to run the update function after input events
+  textarea.addEventListener('input', updateElasticTextArea);
+
+  // Fire an input event to set the initial size correctly
+  textarea.dispatchEvent(new Event('input'));
+
+  // Return an object with the destroy method
+  // so users can remove the event listener if needed
+  return {
+    destroy() {
+      textarea.removeEventListener('input', updateElasticTextArea);
+    },
+  };
```

Comment:
Perhaps nothing to be handled in this PR but I was curious what folks think about this:
```js
export const createElasticTextArea = () => {
  return {
    destroy() {
      // some code...
    }
  }
}
```
vs
```js
export const createElasticTextArea = () => {
  return {
    destroy: () => {
      // some code...
    }
  }
}
```
vs
```js
export const createElasticTextArea = () => {
  const destroy = () => {
    // some code
  }

  return {
    destroy
  }
}

```
I _personally_ prefer the second or third but curious how the rest of the team feels about it. 🙂

Reply:
> [I would add a description annotation to the destroy method, which would be helpful:](https://github.com/cloudfour/cloudfour.com-patterns/pull/716#discussion_r428257857)

This is one reason why I would consider the third option. It keeps the `return {}` more readable at a glance. 🙂

Reply:
This is one of those things where because I came to JS after ES6 was already a thing, I think the ES6 shorthand method looks simpler, but I don't feel strongly about it.

If we were picking a style for our JS guide, my personal preference would be 1, 3, 2. I don't think 2 adds anything over 3 besides more verbosity, unless I'm missing something. But I could see a case being made for 3 if you had a lot of stuff in the return object, so you don't clutter it up.

Reply:
I agree with Scott. In general I prefer #1 but sometimes #3 is necessary if you need to have internal access to the method you're returning.

Reply:
Updated with option 3

Reply:
This comment has nothing actionable for this PR, so I'm going to resolve this conversation.

Thank you for your input! After reading, I can be convinced any are fine with a lean toward the third. But that is not a blocker to this PR. 😉